analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/CalculationsShouldNotOverflowBase.cs

```diff
@@ -18,12 +18,82 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
+using System.Numerics;
+using SonarAnalyzer.SymbolicExecution.Constraints;
+
 namespace SonarAnalyzer.SymbolicExecution.Roslyn.RuleChecks;
 
 public abstract class CalculationsShouldNotOverflowBase : SymbolicRuleCheck
 {
     protected const string DiagnosticId = "S3949";
-    protected const string MessageFormat = "This calculation is {0} to overflow the maximum value of '{1}'.";
+    protected const string MessageFormat = "This calculation is {0} to {1} value of '{2}'.";
     private const string MessageLikely = "likely";
     private const string MessageGuaranteed = "guaranteed";
+    private const string MessageUnderflow = "underflow the minimum";
+    private const string MessageOverflow = "overflow the maximum";
+
+    protected override ProgramState PostProcessSimple(SymbolicContext context)
+    {
+        if (context.Operation.Instance is var operation
+            && CanOverflow(context.State, operation)?.Constraint<NumberConstraint>() is { } number
+            && Min(operation.Type) is { } typeMin
+            && Max(operation.Type) is { } typeMax)
+        {
+            if (number.Max < typeMin)
+            {
+                ReportIssue(operation, MessageGuaranteed, MessageUnderflow, typeMax);
+            }
+            else if (number.Min > typeMax)
+            {
+                ReportIssue(operation, MessageGuaranteed, MessageOverflow, typeMax);
+            }
+            else if (number.Max > typeMax)
+            {
+                ReportIssue(operation, MessageLikely, MessageOverflow, typeMax);
+            }
+            else if (number.Min < typeMin)
+            {
+                ReportIssue(operation, MessageLikely, MessageUnderflow, typeMax);
+            }
+        }
+
+        return context.State;
+    }
+
+    private static SymbolicValue CanOverflow(ProgramState state, IOperation operation) => operation.Kind switch
```

Comment:
I initially wanted to return operation, but that's not possible due to the tricky post increment/decrement :/
So let's give it a better name, as `CanOverflow` answer a question that should return `bool`.

```suggestion
    private static SymbolicValue OverflowCandidateValue(ProgramState state, IOperation operation) => operation.Kind switch
```
or
```suggestion
    private static SymbolicValue OverflowValue(ProgramState state, IOperation operation) => operation.Kind switch
```
or something like that